\name{read.starbeast}
\alias{read.starbeast}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
this function reads in the results of a *BEAST phylogenetic analysis}
\description{
%%  ~~ A concise (1-5 lines) description of what the function does. ~~
}
\usage{
read.starbeast(beast.xml, combinedfiledirectory, logfile = "combined.log")
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{beast.xml}{
The path and name of the xml file for the *BEAST analysis to be checked.}
  \item{combinedfiledirectory}{
The location of the species and gene tree files and the log files generated by *BEAST. 
}
  \item{logfile}{
the name of the logfile. 
}
}
\details{
This function reads in the data (via an xml file), the posterior distribution of gene trees, species trees and model parameters (via the log file). It assumes that the tree files have the name mylocus.combined.trees and the log file is called combined.log. This is because it assumes that the files have had the burn-in removed and been thinned (and possibly multiple runs combined) using a perl script "beast.tree.log.combiner.pl". Running that script is recommended because it's a lot easier than doing all that stuff by hand. 

This script takes way too long, primarily because the way that it extracts the among-branch rate variation parameters and the species tree effective population sizes is slow. If anyone has a better way, let Noah Reid know and he'll implement it!
}
\value{
The function returns an object of class "starbeast.data". It is a list that contains several objects:

associations: maps alleles to species

alignments: contains DNA sequence alignments

gene.trees: A list of ape multiphylo objects containing posterior samples of coalescent genealogies for each locus in the analysis. branch rate variation is stored in a vector named "rate" appended to each phylo object. 

species.trees: An ape multiphylo object containing posterior samples of the species tree. branch effective population sizes are stored in a vector named "dmv" appended to each phylo object. 

log: *BEAST log file values for each sample from the posterior. 

genes: A vector of the names for each locus. 

genes.to.partitions: Maps sequence data partitions to their respective loci. 

ploidy: A vector of numbers assigning ploidy to each locus. All values are 1 if each locus has the same ploidy. If ploidy varies, organellar DNA is 0.5, nuclear DNA is 2.0. 
}
\references{
Reid et al. 2012
Heled and Drummond 2010
}
\author{
Noah Reid
}
\note{
%%  ~~further notes~~
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{
##---- Should be DIRECTLY executable !! ----
##-- ==>  Define data, use random,
##--	or do  help(data=index)  for the standard data sets.

}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.
\keyword{ ~kwd1 }
\keyword{ ~kwd2 }% __ONLY ONE__ keyword per line
